File: sem2/2a.py
n, m, a = map(int, input().split())

def genA(prev):
    return (23 * prev + 21563) % 16714589

def genUV(ind, p_u, p_v, p_r):
    return ((17 * p_u + 751 + p_r + 2 * ind) % n) + 1, \
           ((13 * p_v + 593 + p_r + 5 * ind) % n) + 1

st = [[999999999] * (len(bin(n + 1)) - 2) for _ in range(n)]
st[0][0] = a
for i in range(1, n):
    st[i][0] = genA(st[i - 1][0])

for j in range(1, len(st[0])):
    for i in range(n):
        st[i][j] = st[i][j - 1]
        if i + 2 ** (j - 1) < n:
            st[i][j] = min(st[i][j], st[i + 2 ** (j - 1)][j - 1])

u, v = map(int, input().split())

for i in range(1, m + 1):
    x, y = min(u, v), max(u, v)

    j = len(bin(y - x + 1)) - 3
    r = min(st[x - 1][j], st[y - 2 ** j][j])

    if i < m:
        u, v = genUV(i, u, v, r)

print(u, v, r)

File: sem2/2a.cpp
#include <bits/stdc++.h>

using namespace std;

long long n;

long long genA(long long a) {
    return (23 * a + 21563) % 16714589;
}

long long genU(long long i, long long u, long long r) {
    return ((17 * u + 751 + r + 2 * i) % n) + 1;
}

long long genV(long long i, long long v, long long r) {
    return ((13 * v + 593 + r + 5 * i) % n) + 1;
}

int main() {
    long long m, u, v, r, k;
    cin >> n >> m;
    k = log(n) / log(2) + 1;
    long long st[n][k];
    cin >> st[0][0];

    for (long long i = 1; i < n; ++i) {
        st[i][0] = genA(st[i - 1][0]);
    }

    for (long long j = 1; j < k; ++j) {
        for (long long i = 0; i < n; ++i) {
            st[i][j] = st[i][j - 1];
            if (1 << (j - 1) < n - i) {
                st[i][j] = min(st[i][j], st[i + (1 << (j - 1))][j - 1]);
            }
        }
    }

    cin >> u >> v;

    for (long long i = 1; i <= m; ++i) {
        long long x = min(u, v);
        long long y = max(u, v);

        long long j = log(y - x + 1) / log(2);
        r = min(st[x - 1][j], st[y - (1 << j)][j]);

        if (i < m) {
            u = genU(i, u, r);
            v = genV(i, v, r);
        }
    }

    cout << u << ' ' << v << ' ' << r;
}